src/main/java/org/tikv/common/region/RegionStoreClient.java

```diff
@@ -427,13 +427,367 @@ private void handleRawBatchPut(RawBatchPutResponse resp) {
+          bo.doBackOff(BoTxnLockFast, new KeyException((resp.getError().getLocked().toString())));
+        }
+        return false;
+      } else {
+        throw new KeyException(resp.getError());
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Delete all expired history kv data which version less than specific safePoint
+   * @param bo
+   * @param safePoint
+   */
+  public void gc(BackOffer bo, long safePoint){
+    while(true) {
+      Supplier<Kvrpcpb.GCRequest> factory =
+              () -> Kvrpcpb.GCRequest.newBuilder().setSafePoint(safePoint).setContext(region.getContext()).build();
+      KVErrorHandler<Kvrpcpb.GCResponse> handler =
+              new KVErrorHandler<>(
+                      regionManager,
+                      this,
+                      region,
+                      resp -> resp.hasRegionError() ? resp.getRegionError() : null);
+      Kvrpcpb.GCResponse resp = callWithRetry(bo, TikvGrpc.METHOD_KV_GC, factory, handler);
+      if (gcHelper(bo, resp)) {
+        break;
+      }
+    }
+  }
+
+  private boolean gcHelper(BackOffer bo, Kvrpcpb.GCResponse resp) {
+    if(resp == null){
+      this.regionManager.onRequestFail(region);
+      throw new TiClientInternalException("GCResponse failed without a cause");
+    }
+    if(resp.hasRegionError()){
+      bo.doBackOff(BoRegionMiss, new RegionException(resp.getRegionError()));
+      return false;
+    }
+    if(resp.hasError()) {
+      if (resp.getError().hasLocked()) {
+        Lock lock = new Lock(resp.getError().getLocked());
+        boolean ok = lockResolverClient.resolveLocks(bo, new ArrayList<>(Arrays.asList(lock)));
+        if (!ok) {
+          bo.doBackOff(BoTxnLockFast, new KeyException((resp.getError().getLocked().toString())));
+        }
+        return false;
+      } else {
+        throw new KeyException(resp.getError());
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Scan all locks which version less than a specific version
+   * @param bo
+   * @param startKey
+   * @param maxVersion
+   * @param limit
+   * @return
+   */
+  public List<Kvrpcpb.LockInfo> scanLock(BackOffer bo, ByteString startKey , long maxVersion, int limit){
+    while(true) {
+      Supplier<Kvrpcpb.ScanLockRequest> factory =
+              () -> Kvrpcpb.ScanLockRequest.newBuilder().setContext(region.getContext())
+                      .setMaxVersion(maxVersion).setStartKey(startKey).setLimit(limit).build();
+      KVErrorHandler<Kvrpcpb.ScanLockResponse> handler =
+              new KVErrorHandler<>(
+                      regionManager,
+                      this,
+                      region,
+                      resp -> resp.hasRegionError() ? resp.getRegionError() : null);
+      Kvrpcpb.ScanLockResponse resp = callWithRetry(bo, TikvGrpc.METHOD_KV_SCAN_LOCK, factory, handler);
+      if (scanLockHelper(bo, resp)) {
+        return resp.getLocksList();
+      }
+
+      // we should refresh region
+      region = regionManager.getRegionByKey(startKey);
```

Comment:
why refresh region here?

Reply:
When encountering an region error, we should request for region info to PD.